packages/kbn-alerts-grouping/src/alerts_grouping/components/alerts_grouping.tsx

```diff
@@ -0,0 +1,260 @@
+    Array(MAX_GROUPING_LEVELS).fill(DEFAULT_PAGE_INDEX)
+  );
+  const [pageSize, setPageSize] = useState<number[]>(getStoragePageSize);
+
+  const resetAllPagination = useCallback(() => {
+    setPageIndex((curr) => curr.map(() => DEFAULT_PAGE_INDEX));
+  }, []);
+
+  const setPageVar = useCallback(
+    (newNumber: number, groupingLevel: number, pageType: 'index' | 'size') => {
+      if (pageType === 'index') {
+        setPageIndex((currentIndex) => {
+          const newArr = [...currentIndex];
+          newArr[groupingLevel] = newNumber;
+          return newArr;
+        });
+      }
+
+      if (pageType === 'size') {
+        setPageSize((currentIndex) => {
+          const newArr = [...currentIndex];
+          newArr[groupingLevel] = newNumber;
+          setStoragePageSize(newArr);
+          return newArr;
+        });
+        // set page index to 0 when page size is changed
+        setPageIndex((currentIndex) => {
+          const newArr = [...currentIndex];
+          newArr[groupingLevel] = 0;
+          return newArr;
+        });
+      }
+    },
+    [setStoragePageSize]
+  );
+
+  const paginationResetTriggers = useRef({
+    defaultFilters,
+    globalFilters,
+    globalQuery,
+    selectedGroups,
+  });
+
+  useEffect(() => {
+    const triggers = {
+      defaultFilters,
+      globalFilters,
+      globalQuery,
+      selectedGroups,
+    };
+    if (!isEqual(paginationResetTriggers.current, triggers)) {
+      resetAllPagination();
+      paginationResetTriggers.current = triggers;
+    }
+  }, [defaultFilters, globalFilters, globalQuery, resetAllPagination, selectedGroups]);
+
+  const getLevel = useCallback(
+    (level: number, selectedGroup: string, parentGroupingFilter?: string) => {
+      const resetGroupChildrenPagination = (parentLevel: number) => {
+        setPageIndex((allPages) => {
+          const resetPages = allPages.splice(parentLevel + 1, allPages.length);
+          return [...allPages, ...resetPages.map(() => DEFAULT_PAGE_INDEX)];
+        });
+      };
+
+      return (
+        <AlertsGroupingLevel
+          {...props}
+          getGrouping={getGrouping}
+          groupingLevel={level}
+          onGroupClose={() => resetGroupChildrenPagination(level)}
+          pageIndex={pageIndex[level] ?? DEFAULT_PAGE_INDEX}
+          pageSize={pageSize[level] ?? DEFAULT_PAGE_SIZE}
+          parentGroupingFilter={parentGroupingFilter}
+          selectedGroup={selectedGroup}
+          setPageIndex={(newIndex: number) => setPageVar(newIndex, level, 'index')}
+          setPageSize={(newSize: number) => setPageVar(newSize, level, 'size')}
+        >
+          {level < selectedGroups.length - 1
+            ? (groupingFilters: Filter[]) => {
+                return getLevel(
+                  level + 1,
+                  selectedGroups[level + 1],
+                  // stringify because if the filter is passed as an object, it will cause unnecessary re-rendering
```

Comment:
Could memoization help us to avoid `JSON.stringify` and `JSON.parse` which are a bit expensive operations?

Reply:
I tried a million different combinations (the tricky part was the recursive function and `{...props}` forwarding causing again unnecessary renders. [This](https://github.com/elastic/kibana/pull/184635/commits/95cdfbc33a33bf511555ce089b34ebbe7047cdf2) seems to work pretty well!

Reply:
TIL! Amazing solution! It reminded this great article https://kentcdodds.com/blog/optimize-react-re-renders.